api/src/modules/Session/model.ts

```diff
@@ -74,6 +82,51 @@ const generateTokenAndSession = async (
   };
 };
 
+async function createUserIfNotExists(prisma: PrismaClient, wallet_address: string) {
+  const user = await prisma.user.findUnique({
+    where: {
+      id: wallet_address,
+    }
+  })
+  if (!!user) {
+    return user
+  }
+  return await prisma.user.create({
+    data: {
+      id: wallet_address,
+    }
+  })
+
+}
+async function createPendingSession(prisma: PrismaClient): Promise<string> {
+  const id = randomUUID();
+  await prisma.pendingAuth.create({
+    data: {
+      id,
+    },
+  });
+  return id
+}
+
+async function verifyMessageAndCreateSession(
+  prisma: PrismaClient,
+  message: string,
+  signature: string,
+  session: { expiryDurationSeconds?: number | null; name: string },
+  isUserCreated: boolean = false
+) {
+  const siweMessage = new SiweMessage(message);
+  const userId = siweMessage.address;
+  const sessionId = siweMessage.nonce;
```

Comment:
The whole point of `nonce` is to validate that it exists in our database, but currently you're not doing that. As you don't delete used challenges from our database the replay attack would still be possible, even if you would validate it. Moreover you're misleading by calling it `sessionId`, since it's not yet a session, it's a challengeId.

- [ ] Make sure that this function throws error if nonce is unknown to us
- [ ] Make sure to delete the challenge once it used once
- [ ] Make sure to do everything in one sql transaction